pkg/controllers/dynakube/extension/tls/reconciler.go

```diff
@@ -0,0 +1,117 @@
+package tls
+
+import (
+	"context"
+	"crypto/x509"
+
+	"github.com/Dynatrace/dynatrace-operator/pkg/api/v1beta3/dynakube"
+	"github.com/Dynatrace/dynatrace-operator/pkg/controllers"
+	"github.com/Dynatrace/dynatrace-operator/pkg/controllers/dynakube/extension/consts"
+	"github.com/Dynatrace/dynatrace-operator/pkg/util/certificates"
+	k8slabels "github.com/Dynatrace/dynatrace-operator/pkg/util/kubeobjects/labels"
+	k8ssecret "github.com/Dynatrace/dynatrace-operator/pkg/util/kubeobjects/secret"
+	"github.com/Dynatrace/dynatrace-operator/pkg/util/timeprovider"
+	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type reconciler struct {
+	client       client.Client
+	apiReader    client.Reader
+	timeProvider *timeprovider.Provider
+
+	dk *dynakube.DynaKube
+}
+
+func NewReconciler(clt client.Client, apiReader client.Reader, dk *dynakube.DynaKube) controllers.Reconciler {
+	return &reconciler{
+		client:       clt,
+		apiReader:    apiReader,
+		dk:           dk,
+		timeProvider: timeprovider.New(),
+	}
+}
+
+func (r *reconciler) Reconcile(ctx context.Context) error {
+	query := k8ssecret.Query(r.client, r.client, log)
+
+	if !r.dk.ExtensionsNeedsSelfSignedTLS() {
+		return query.Delete(ctx, &corev1.Secret{ObjectMeta: v1.ObjectMeta{Name: getSelfSignedTLSSecretName(r.dk.Name), Namespace: r.dk.Namespace}})
+	}
+
+	_, err := query.Get(ctx, types.NamespacedName{
+		Name:      getSelfSignedTLSSecretName(r.dk.Name),
+		Namespace: r.dk.Namespace,
+	})
+
+	if err != nil && !k8serrors.IsNotFound(err) {
+		return err
+	}
+
+	if k8serrors.IsNotFound(err) {
+		err = r.createOrUpdateTLSSecret(ctx)
+		if err != nil {
+			return err
+		}
+
+		return nil
+	}
```

Comment:
if I understand correctly, the scenario of "update" will never happen, only create

because `k8serrors.IsNotFound(err)` will be `false` if the secret already exists, so you will just `return nil`

you don't need that if
```suggestion
	if err != nil && !k8serrors.IsNotFound(err) {
		return err
	}

	err = r.createOrUpdateTLSSecret(ctx)
	if err != nil {
		return err
	}
```

Reply:
[Fixed as discussed in Slack](https://github.com/Dynatrace/dynatrace-operator/pull/3778/commits/687ec32290257b5e583fafaead823bc9e427c06a)